src/preprocessing/passes/daneshvar.cpp

```diff
@@ -0,0 +1,817 @@
+        return 0;
+    if (k == cvc5::internal::Kind::XOR)
+        return 0;
+    if (k == cvc5::internal::Kind::DISTINCT)
+        return 0;
+    if (k == cvc5::internal::Kind::EQUAL) 
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_AND)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_OR)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_XOR)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_NAND)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_NOR)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_COMP)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_ADD)
+        return 0;
+    if (k == cvc5::internal::Kind::BITVECTOR_MULT)
+        return 0;
+    if (k == cvc5::internal::Kind::FLOATINGPOINT_EQ)
+        return 0;
+    if (k == cvc5::internal::Kind::FLOATINGPOINT_ADD)
+        return 1;
+    if (k == cvc5::internal::Kind::FLOATINGPOINT_MULT)
+        return 1;
+    return -1;
+}
+
+bool sameClass(NodeInfo a, NodeInfo b)
+{
+    if (a.encoding != b.encoding)
+    {
+        return false;
+    }
+    int n = a.pat.size();
+    for (int i = 0; i < n; i++)
+    {
+        if (a.pat[i] != b.pat[i])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+
+Node sortOperands(std::map<int, std::vector<NodeInfo>>& ec_ass,
+                  std::map<int, std::vector<NodeInfo>>& ec_oper,
+                  const NodeInfo& ni)
+{
+    Node n = ni.node;
+    if (n.isVar() || n.isConst()) {
+        return n;
+    }
+
+    std::stack<NodeInfo> toVisit;
+
+    std::map<Node, std::vector<NodeInfo>> nodeChildrenMap;
+
+    std::map<Node, Node> processedNodes;
+
+    toVisit.push(ni);
+
+    while (!toVisit.empty()) {
+        NodeInfo currentNodeInfo = toVisit.top();
+        Node currentNode = currentNodeInfo.node;
+
+        if (processedNodes.find(currentNode) != processedNodes.end()) {
+            toVisit.pop();
+            continue;
+        }
+
+        if (currentNode.isVar() || currentNode.isConst()) {
+            processedNodes[currentNode] = currentNode;
+            toVisit.pop();
+            continue;
+        }
+
+        if (nodeChildrenMap.find(currentNode) == nodeChildrenMap.end()) {
+            std::vector<NodeInfo> child;
+            for (size_t i = 0; i < currentNode.getNumChildren(); i++) {
+                NodeInfo childInfo = getNodeInfo(currentNode[i], -1, -1);
+                child.push_back(childInfo);
+                toVisit.push(childInfo);
+            }
+            nodeChildrenMap[currentNode] = child;
+
+            if (currentNode.getKind() == cvc5::internal::Kind::APPLY_UF) {
+                NodeInfo operatorInfo = getNodeInfo(currentNode.getOperator(), -1, -1);
+                toVisit.push(operatorInfo);
+            }
+            continue;
+        }
+
+        toVisit.pop();
+
+        std::vector<NodeInfo> child = nodeChildrenMap[currentNode];
```

Comment:
You should avoid storing a copy of a NodeInfo object in the vector. You can use  std::vector<std::reference_wrapper<NodeInfo>>, which will store references to NodeInfo instead of copies.